topic-operator/src/main/java/io/strimzi/operator/topic/v2/BatchingTopicController.java

```diff
@@ -473,15 +473,91 @@ private void updateInternal(List<ReconcilableTopic> topics) {
+                    && reconcilableTopic.kt().getStatus().getReplicasChange().getMessage() != null) {
+                    // the user reverted an invalid replicas value
+                    reconcilableTopic.kt().getStatus().setReplicasChange(null);
+                }
+            });
+        }
+    }
+
+    /* test */ void maybeCheckReplicasChanges(List<ReconcilableTopic> topics,
+                                              Map<ReconcilableTopic, Either<TopicOperatorException, Object>> results,
+                                              PartitionedByError<ReconcilableTopic, CurrentState> differentRfResults) {
+        if (config.cruiseControlEnabled()) {
+            var topicsWithDifferentRf = differentRfResults.ok().map(Pair::getKey).collect(Collectors.toList());
+            requestPendingReplicasChanges(topicsWithDifferentRf, results);
+            checkOngoingReplicasChanges(topics, results);
+        }
+    }
+
+    private void requestPendingReplicasChanges(List<ReconcilableTopic> topicsWithDifferentRf, Map<ReconcilableTopic, Either<TopicOperatorException, Object>> results) {
+        var topicsWithPendingRfChange = topicsMatching(topicsWithDifferentRf, this::isPendingReplicasChange);
+        warnTooLargeMinIsr(topicsWithPendingRfChange);
+        var pendingAndOngoing = replicasChangeClient.requestPendingChanges(topicsWithPendingRfChange);
+        pendingAndOngoing.forEach(rt -> putResult(results, rt, Either.ofRight(null)));
+    }
+
+    private void checkOngoingReplicasChanges(List<ReconcilableTopic> topics, Map<ReconcilableTopic, Either<TopicOperatorException, Object>> results) {
+        var topicsWithOngoingRfChange = topicsMatching(topics, this::isOngoingReplicasChange);
+        var completedAndFailed = replicasChangeClient.requestOngoingChanges(topicsWithOngoingRfChange);
+        completedAndFailed.forEach(rt -> putResult(results, rt, Either.ofRight(null)));
+    }
+
+    private List<ReconcilableTopic> topicsMatching(List<ReconcilableTopic> reconcilableTopics, Predicate<KafkaTopic> status) {
+        return reconcilableTopics.stream().filter(rt -> status.test(rt.kt())).collect(Collectors.toList());
+    }
+
+    private boolean isPendingReplicasChange(KafkaTopic kafkaTopic) {
+        return kafkaTopic.getStatus() != null
+            && (kafkaTopic.getStatus().getReplicasChange() == null
+            || (kafkaTopic.getStatus().getReplicasChange().getState() == PENDING
+            && kafkaTopic.getStatus().getReplicasChange().getSessionId() == null));
+    }
+
+    private boolean isOngoingReplicasChange(KafkaTopic kafkaTopic) {
+        return hasReplicasChange(kafkaTopic.getStatus())
+            && kafkaTopic.getStatus().getReplicasChange().getState() == ONGOING
+            && kafkaTopic.getStatus().getReplicasChange().getSessionId() != null;
+    }
+
+    private void warnTooLargeMinIsr(List<ReconcilableTopic> reconcilableTopics) {
```

Comment:
I see plenty of helper methods here...but this one I think deserves a JavaDoc because of complexity...

Reply:
Ok, done.